pkg/config/config_test.go

```diff
@@ -632,6 +634,140 @@ func TestDefaultGeneralEntries(t *testing.T) {
+	topLevelKey := []string{"toplevelkey"}
+	newHostKey := []string{"hosts", "newhost"}
+
+	migrator := &MigrationMock{
+		DoFunc: func(config *Config) (bool, error) {
+			config.Set(topLevelKey, "toplevelvalue")
+			config.Set(newHostKey, "newhostvalue")
+			return true, nil
+		},
+	}
+
+	// When we run the migration
+	require.NoError(t, Migrate(cfg, migrator))
+
+	// Then our original config is updated with the migration applied
+	requireKeyWithValue(t, cfg, topLevelKey, "toplevelvalue")
+	requireKeyWithValue(t, cfg, newHostKey, "newhostvalue")
+
+	// And our config / hosts changes are persisted to their relevant files
+	persistedCfg, err := load(generalConfigFile(), hostsConfigFile())
+	require.NoError(t, err)
+	requireKeyWithValue(t, persistedCfg, topLevelKey, "toplevelvalue")
+	requireKeyWithValue(t, persistedCfg, newHostKey, "newhostvalue")
+
+	// And we have a backup file with the contents of the original config
+	backupCfg, err := load(filepath.Join(tempDir, "config.yml.bak"), filepath.Join(tempDir, "hosts.yml.bak"))
+	require.NoError(t, err)
+	requireNoKey(t, backupCfg, topLevelKey)
+	requireNoKey(t, backupCfg, newHostKey)
+}
+
+func TestMigrationNotRequiredWritesNoFiles(t *testing.T) {
+	// Create a location for our tests to write,
+	// Note that using env var here makes these tests unparallelisable.
+	tempDir := t.TempDir()
+	t.Setenv("GH_CONFIG_DIR", tempDir)
+
+	// Given we have a migrator that returns that no migration is required
+	cfg := ReadFromString(testFullConfig())
+	migrator := &MigrationMock{
+		DoFunc: func(config *Config) (bool, error) {
+			return false, nil
+		},
+	}
+
+	// When we run the migration
+	require.NoError(t, Migrate(cfg, migrator))
+
+	// Then no files are written to disk
+	files, err := os.ReadDir(tempDir)
+	require.NoError(t, err)
+	require.Len(t, files, 0)
+}
+
+func TestMigrationErrorWritesNoFiles(t *testing.T) {
+	// Create a location for our tests to write,
+	// Note that using env var here makes these tests unparallelisable.
+	tempDir := t.TempDir()
+	t.Setenv("GH_CONFIG_DIR", tempDir)
+
+	// Given we have a migrator that errors
+	cfg := ReadFromString(testFullConfig())
+	migrator := &MigrationMock{
+		DoFunc: func(config *Config) (bool, error) {
+			return true, errors.New("failed to migrate in test")
+		},
+	}
+
+	// When we run the migration
+	err := Migrate(cfg, migrator)
+
+	// Then the error is wrapped and bubbled
+	require.EqualError(t, err, "failed to migrate config: failed to migrate in test")
+
+	// And no files are written to disk
+	files, err := os.ReadDir(tempDir)
+	require.NoError(t, err)
+	require.Len(t, files, 0)
+}
+
+func TestMigrationBackupWriteError(t *testing.T) {
+	// Create a location for our tests to write,
+	// Note that using env var here makes these tests unparallelisable.
+	tempDir := t.TempDir()
+	t.Setenv("GH_CONFIG_DIR", tempDir)
+
+	// Given we error when writing files (because we explicitly disallow it via this chmod)
+	require.NoError(t, os.Chmod(tempDir, 0000))
```

Comment:
This doesn't work on Windows apparently. Having an interface where we could inject faults on persistence operations would be handy but I think maybe it's just a lot of extra abstraction for limited gain right now.

Reply:
It is clear. I'm generally not a fan of testing the internals like this but since our tests already happen to live in the same package (rather than `_test`) maybe this is a reasonable approach. I'll noodle on it thanks.

Reply:
Doing this approach didn't allow me control over the specific error paths (for hosts vs config) so I went with an integrated solution of making the files unwriteable.